single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,751 @@
+        super(OperatorProtocol, self).__init__(charm, self.name)
+        self.charm = charm
+        self.substrate: Substrates = self.charm.substrate
+        self.role = ROLES[self.substrate][self.name]
+        self.state = CharmState(
+            self.charm,
+            self.substrate,
+            self.role,
+        )
+
+        container = (
+            self.charm.unit.get_container(CONTAINER) if self.substrate == Substrates.K8S else None
+        )
+
+        # Defined workloads and configs
+        self.define_workloads_and_config_managers(container)
+
+        # Managers
+        self.backup_manager = BackupManager(
+            self.charm,
+            self.role,
+            self.substrate,
+            self.state,
+            container,
+        )
+        self.tls_manager = TLSManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+        self.mongo_manager = MongoManager(
+            self,
+            self.workload,
+            self.state,
+            self.substrate,
+        )
+
+        # Event Handlers
+        self.password_actions = PasswordActionEvents(self)
+        self.backup_events = BackupEventsHandler(self)
+        self.tls_events = TLSEventsHandler(self)
+        self.primary_events = PrimaryActionHandler(self)
+        self.client_events = DatabaseEventsHandler(self, RelationNames.DATABASE)
+
+    @property
+    def config(self):
+        """Returns the actual config."""
+        return self.charm.parsed_config
+
+    def define_workloads_and_config_managers(self, container: Container | None) -> None:
+        """Export all workload and config definition for readability."""
+        # BEGIN: Define workloads.
+        self.workload = get_mongodb_workload_for_substrate(self.substrate)(
+            role=self.role, container=container
+        )
+        self.mongos_workload = get_mongos_workload_for_substrate(self.substrate)(
+            role=self.role, container=container
+        )
+        # END: Define workloads
+
+        # BEGIN Define config managers
+        self.config_manager = MongoDBConfigManager(
+            self.config,
+            self.state,
+            self.workload,
+        )
+        self.mongos_config_manager = MongosConfigManager(
+            self.config,
+            self.mongos_workload,
+            self.state,
+        )
+        self.logrotate_config_manager = LogRotateConfigManager(
+            self.role,
+            self.substrate,
+            self.config,
+            self.state,
+            container,
+        )
+        self.mongodb_exporter_config_manager = MongoDBExporterConfigManager(
+            self.role,
+            self.substrate,
+            self.config,
+            self.state,
+            container,
+        )
+        # END: Define config managers
+
+    @property
+    def config_managers(self) -> Iterable[CommonConfigManager]:  # pragma: nocover
```

Comment:
They are config managers because they manage the service config.
But this method goes away in another PR, it was used only in one case and incorrectly so I dropped it.